File: app/src/main/java/com/example/taxconnect/services/CloudinaryHelper.java
package com.example.taxconnect.services;

import android.content.Context;
import android.net.Uri;
import android.util.Log;
import android.webkit.MimeTypeMap;

import com.cloudinary.android.MediaManager;
import com.cloudinary.android.callback.ErrorInfo;
import com.cloudinary.android.callback.UploadCallback;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

public class CloudinaryHelper {
    private static final String TAG = "CloudinaryHelper";
    // Replace with your actual Cloudinary credentials
    private static final String CLOUD_NAME = com.example.taxconnect.BuildConfig.CLOUDINARY_CLOUD_NAME;
    private static final String UPLOAD_PRESET = com.example.taxconnect.BuildConfig.CLOUDINARY_UPLOAD_PRESET;

    public static void init(Context context) {
        try {
            Map<String, Object> config = new HashMap<>();
            config.put("cloud_name", CLOUD_NAME);
            config.put("secure", true);
            MediaManager.init(context, config);
        } catch (IllegalStateException e) {
            // Already initialized
            Log.w(TAG, "Cloudinary already initialized");
        }
    }

    public interface ImageUploadCallback {
        void onSuccess(String url);
        void onError(String error);
    }

    public static void uploadMedia(Context context, Uri uri, ImageUploadCallback callback) {
        // Copy to cache to ensure access and valid file path
        String filePath = copyUriToCache(context, uri);
        
        if (filePath == null) {
            callback.onError("Failed to process file for upload.");
            return;
        }

        MediaManager.get().upload(filePath)
                .unsigned(UPLOAD_PRESET)
                .option("resource_type", "auto") // Handle both image and video
                .callback(new UploadCallback() {
                    @Override
                    public void onStart(String requestId) {
                        Log.d(TAG, "Upload started: " + requestId);
                    }

                    @Override
                    public void onProgress(String requestId, long bytes, long totalBytes) {
                        // Optional: Update progress
                    }

                    @Override
                    @SuppressWarnings({"unchecked", "rawtypes"})
                    public void onSuccess(String requestId, Map resultData) {
                        String url = (String) resultData.get("secure_url");
                        Log.d(TAG, "Upload success: " + url);
                        
                        // Clean up temp file
                        new File(filePath).delete();
                        
                        if (callback != null) {
                            callback.onSuccess(url);
                        }
                    }

                    @Override
                    public void onError(String requestId, ErrorInfo error) {
                        Log.e(TAG, "Upload error: " + error.getDescription());
                        
                        // Clean up temp file
                        new File(filePath).delete();
                        
                        if (callback != null) {
                            callback.onError(error.getDescription());
                        }
                    }

                    @Override
                    public void onReschedule(String requestId, ErrorInfo error) {
                        // Handle reschedule
                    }
                })
                .dispatch();
    }
    
    private static String copyUriToCache(Context context, Uri uri) {
        try {
            InputStream inputStream = context.getContentResolver().openInputStream(uri);
            if (inputStream == null) return null;

            // Try to determine extension
            String extension = ".tmp";
            String mimeType = context.getContentResolver().getType(uri);
            if (mimeType != null) {
                String ext = MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType);
                if (ext != null) extension = "." + ext;
            }

            File tempFile = File.createTempFile("upload_", extension, context.getCacheDir());
            FileOutputStream outputStream = new FileOutputStream(tempFile);

            byte[] buffer = new byte[4096];
            int bytesRead;
            while ((bytesRead = inputStream.read(buffer)) != -1) {
                outputStream.write(buffer, 0, bytesRead);
            }
            
            outputStream.close();
            inputStream.close();
            
            return tempFile.getAbsolutePath();
        } catch (IOException e) {
            Log.e(TAG, "Error copying file to cache: " + e.getMessage());
            return null;
        }
    }

    public static void uploadImage(String filePath, ImageUploadCallback callback) {
        MediaManager.get().upload(filePath)
                .unsigned(UPLOAD_PRESET)
                .option("resource_type", "auto")
                .callback(new UploadCallback() {
                    @Override
                    public void onStart(String requestId) {
                        Log.d(TAG, "Upload started: " + requestId);
                    }

                    @Override
                    public void onProgress(String requestId, long bytes, long totalBytes) {
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public void onSuccess(String requestId, Map resultData) {
                        String url = (String) resultData.get("secure_url");
                        if (callback != null) callback.onSuccess(url);
                    }

                    @Override
                    public void onError(String requestId, ErrorInfo error) {
                        if (callback != null) callback.onError(error.getDescription());
                    }

                    @Override
                    public void onReschedule(String requestId, ErrorInfo error) {
                    }
                })
                .dispatch();
    }
}

File: tax/TAX-connect-prototype/tax/app/src/main/java/com/example/taxconnect/data/services/CloudinaryHelper.kt
package com.example.taxconnect.data.services

import android.content.Context
import android.net.Uri
import android.util.Log
import android.webkit.MimeTypeMap
import com.cloudinary.android.MediaManager
import com.cloudinary.android.callback.ErrorInfo
import com.cloudinary.android.callback.UploadCallback
import com.example.taxconnect.BuildConfig
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

object CloudinaryHelper {
    private const val TAG = "CloudinaryHelper"
    // Replace with your actual Cloudinary credentials
    private const val CLOUD_NAME = BuildConfig.CLOUDINARY_CLOUD_NAME
    private const val UPLOAD_PRESET = BuildConfig.CLOUDINARY_UPLOAD_PRESET

    fun init(context: Context) {
        try {
            val config: MutableMap<String, Any> = HashMap()
            config["cloud_name"] = CLOUD_NAME
            config["secure"] = true
            MediaManager.init(context, config)
        } catch (e: IllegalStateException) {
            // Already initialized
            Log.w(TAG, "Cloudinary already initialized")
        }
    }

    interface ImageUploadCallback {
        fun onSuccess(url: String?)
        fun onError(error: String?)
    }

    fun uploadMedia(context: Context, uri: Uri, callback: ImageUploadCallback?) {
        // Copy to cache to ensure access and valid file path
        val filePath = copyUriToCache(context, uri)

        if (filePath == null) {
            callback?.onError("Failed to process file for upload.")
            return
        }

        MediaManager.get().upload(filePath)
            .unsigned(UPLOAD_PRESET)
            .option("resource_type", "auto") // Handle both image and video
            .callback(object : UploadCallback {
                override fun onStart(requestId: String) {
                    Log.d(TAG, "Upload started: $requestId")
                }

                override fun onProgress(requestId: String, bytes: Long, totalBytes: Long) {
                    // Optional: Update progress
                }

                override fun onSuccess(requestId: String, resultData: Map<*, *>) {
                    val url = resultData["secure_url"] as String?
                    Log.d(TAG, "Upload success: $url")

                    // Clean up temp file
                    File(filePath).delete()

                    callback?.onSuccess(url ?: "")
                }

                override fun onError(requestId: String, error: ErrorInfo) {
                    Log.e(TAG, "Upload error: " + error.description)

                    // Clean up temp file
                    File(filePath).delete()

                    callback?.onError(error.description)
                }

                override fun onReschedule(requestId: String, error: ErrorInfo) {
                    // Handle reschedule
                }
            })
            .dispatch()
    }

    fun uploadFile(filePath: String, callback: ImageUploadCallback?) {
        MediaManager.get().upload(filePath)
            .unsigned(UPLOAD_PRESET)
            .option("resource_type", "auto")
            .callback(object : UploadCallback {
                override fun onStart(requestId: String) {}
                override fun onProgress(requestId: String, bytes: Long, totalBytes: Long) {}
                override fun onSuccess(requestId: String, resultData: Map<*, *>) {
                    callback?.onSuccess(resultData["secure_url"] as String?)
                }
                override fun onError(requestId: String, error: ErrorInfo) {
                    callback?.onError(error.description)
                }
                override fun onReschedule(requestId: String, error: ErrorInfo) {}
            })
            .dispatch()
    }

    private fun copyUriToCache(context: Context, uri: Uri): String? {
        try {
            context.contentResolver.openInputStream(uri).use { inputStream ->
                if (inputStream == null) return null

                // Try to determine extension
                var extension = ".tmp"
                val mimeType = context.contentResolver.getType(uri)
                if (mimeType != null) {
                    val ext = MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType)
                    if (ext != null) extension = ".$ext"
                }

                val tempFile = File.createTempFile("upload_", extension, context.cacheDir)
                FileOutputStream(tempFile).use { outputStream ->
                    val buffer = ByteArray(4096)
                    var bytesRead: Int
                    while (inputStream.read(buffer).also { bytesRead = it } != -1) {
                        outputStream.write(buffer, 0, bytesRead)
                    }
                }
                return tempFile.absolutePath
            }
        } catch (e: IOException) {
            Log.e(TAG, "Error copying file to cache: " + e.message)
            return null
        }
    }

    fun uploadImage(filePath: String, callback: ImageUploadCallback?) {
        MediaManager.get().upload(filePath)
            .unsigned(UPLOAD_PRESET)
            .option("resource_type", "auto")
            .callback(object : UploadCallback {
                override fun onStart(requestId: String) {
                    Log.d(TAG, "Upload started: $requestId")
                }

                override fun onProgress(requestId: String, bytes: Long, totalBytes: Long) {
                }

                override fun onSuccess(requestId: String, resultData: Map<*, *>) {
                    val url = resultData["secure_url"] as String?
                    callback?.onSuccess(url ?: "")
                }

                override fun onError(requestId: String, error: ErrorInfo) {
                    callback?.onError(error.description)
                }

                override fun onReschedule(requestId: String, error: ErrorInfo) {
                }
            })
            .dispatch()
    }
}
